File: C++/148. SortList.cpp
/**
 *Implementation of an N * log (n) algorithm for sorting a linked list, with node as defined below, using MergeSort being used.
 *
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 *
 * For Example:
 * Input: [-1,5,3,4,0]
 * Output: [-1,0,3,4,5]
 */
class Solution {
public:
    void mergeSort(ListNode** headRef) 
    { 
        ListNode* head = *headRef; 
        ListNode* a; 
        ListNode* b; 
  
        /* Base case -- length 0 or 1 */
        if ((head == NULL) || (head->next == NULL)) { 
            return; 
        } 
  
        /* Split head into 'a' and 'b' sublists */
        frontBackSplit(head, &a, &b); 
  
        /* Recursively sort the sublists */
        mergeSort(&a); 
        mergeSort(&b); 
  
         /* answer = merge the two sorted lists together */
        *headRef = sortedMerge(a, b); 
    } 
    
    void frontBackSplit(ListNode* source, ListNode** frontRef, ListNode** backRef) 
    { 
        ListNode* fast; 
        ListNode* slow; 
        slow = source; 
        fast = source->next; 
  
        /* Advance 'fast' two nodes, and advance 'slow' one node */
        while (fast != NULL) { 
            fast = fast->next; 
            if (fast != NULL) { 
                slow = slow->next; 
                fast = fast->next; 
            }   
        } 
  
        /* 'slow' is before the midpoint in the list, so split it in two  
        at that point. */
        *frontRef = source; 
        *backRef = slow->next; 
        slow->next = NULL; 
    } 
    
    ListNode* sortedMerge(ListNode* a, ListNode* b) 
    { 
        ListNode* result = NULL; 
  
        /* Base cases */
        if (a == NULL) 
            return (b); 
        else if (b == NULL) 
            return (a); 
  
        /* Pick either a or b, and recur */
        if (a->val <= b->val) { 
            result = a; 
            result->next = sortedMerge(a->next, b); 
        } 
        else { 
            result = b; 
            result->next = sortedMerge(a, b->next); 
        } 
        return (result); 
    } 
    
    ListNode* sortList(ListNode* head) {
        mergeSort(&head);
        return head;
    }
};

File: Python/148. SortList.py
#Problem: Given the head of a linked list, return the list after sorting it in ascending order.
#Approach: We use merge sort and account for the use of a linked list to solve this

class Solution(object):
    #returns the middle node of the linked list referenced by head
    def getMiddle(self,head):
        double_step = head
        single_step = head
        prev_single = head
        while (double_step is not None) and (double_step.next is not None):
            prev_single = single_step
            single_step = single_step.next
            double_step = double_step.next.next
        prev_single.next = None
        return single_step    

    #merges in sorted order two sorted linked lists
    def merge(self,head_of_list1, head_of_list2):
        temp = ListNode(0);
        temp_head = temp;
        while head_of_list1 and head_of_list2:
            if head_of_list1.val < head_of_list2.val:
                temp.next = head_of_list1
                head_of_list1 = head_of_list1.next
            else:
                temp.next = head_of_list2
                head_of_list2 = head_of_list2.next
            temp = temp.next
        if head_of_list1:
            temp.next = head_of_list1
        elif head_of_list2:
            temp.next = head_of_list2
        return temp_head.next
        """
        :type head: ListNode
        :rtype: ListNode
        """
    def sortList(self, head):
        if (head is None) or (head.next is None):
            return head
        middle = self.getMiddle(head)
        left = self.sortList(head)
        right = self.sortList(middle)
        return self.merge(left, right)
